File: apps/android/app/src/main/java/com/tonezen/app/domain/downloads/AudiobookDownloadTarget.kt
package com.tonezen.app.domain.downloads

import com.tonezen.app.domain.model.Book
import com.tonezen.app.domain.model.Track

fun nextAudiobookDownloadRequest(
    book: Book,
    tracks: List<Track>,
    currentTrackId: String?,
    savedTrackId: String?,
): EnqueueDownloadRequest? {
    val track = nextAudiobookDownloadTrack(tracks, currentTrackId, savedTrackId) ?: return null
    return EnqueueDownloadRequest(
        bookId = book.id,
        trackId = track.id,
        priority = DownloadPriority.USER,
        title = track.title,
        subtitle = book.title,
        contentType = book.contentType.name.lowercase(),
    )
}

fun nextAudiobookDownloadTrack(
    tracks: List<Track>,
    currentTrackId: String?,
    savedTrackId: String?,
): Track? {
    val sorted = tracks.sortedBy { it.sortOrder }
    if (sorted.isEmpty()) return null

    val currentIndex = currentTrackId?.let { id -> sorted.indexOfFirst { it.id == id } } ?: -1
    if (currentIndex >= 0) {
        return sorted.drop(currentIndex + 1).firstOrNull { it.localPath.isNullOrBlank() }
    }

    val savedIndex = savedTrackId?.let { id -> sorted.indexOfFirst { it.id == id } } ?: -1
    val startIndex = if (savedIndex >= 0) savedIndex else 0
    return sorted.drop(startIndex).firstOrNull { it.localPath.isNullOrBlank() }
}

File: apps/desktop/src/core/downloads/audiobookDownloadTarget.ts
import type { EnqueueDownloadRequest } from "@core/downloads/downloadQueueState.js";
import type { Book, Track } from "@core/types.js";

interface NextAudiobookDownloadInput {
  book: Book;
  tracks: Track[];
  currentTrackId: string | null;
  savedTrackId: string | null;
}

export function nextAudiobookDownloadRequest({
  book,
  tracks,
  currentTrackId,
  savedTrackId,
}: NextAudiobookDownloadInput): EnqueueDownloadRequest | null {
  const track = nextAudiobookDownloadTrack(tracks, currentTrackId, savedTrackId);
  if (!track) return null;
  return {
    bookId: book.id,
    trackId: track.id,
    priority: "USER",
    title: track.title,
    subtitle: book.title,
    contentType: book.contentType,
  };
}

export function nextChapterInBook(tracks: Track[], currentTrackId: string): Track | null {
  const sorted = [...tracks].sort((a, b) => a.sortOrder - b.sortOrder);
  const index = sorted.findIndex((track) => track.id === currentTrackId);
  if (index < 0 || index >= sorted.length - 1) return null;
  return sorted[index + 1];
}

function nextAudiobookDownloadTrack(
  tracks: Track[],
  currentTrackId: string | null,
  savedTrackId: string | null,
): Track | null {
  const sorted = [...tracks].sort((a, b) => a.sortOrder - b.sortOrder);
  if (sorted.length === 0) return null;

  const currentIndex = currentTrackId
    ? sorted.findIndex((track) => track.id === currentTrackId)
    : -1;
  if (currentIndex >= 0) {
    return sorted.slice(currentIndex + 1).find((track) => !track.localPath) ?? null;
  }

  const savedIndex = savedTrackId ? sorted.findIndex((track) => track.id === savedTrackId) : -1;
  const startIndex = savedIndex >= 0 ? savedIndex : 0;
  return sorted.slice(startIndex).find((track) => !track.localPath) ?? null;
}
